### DISCUSSION TITLE
Flipper Icon?

### DISCUSSION THREAD

**Discussion by GitCat3** [dcoles/flipperzero-rs#20] [Q&A] [answered]
So I notice that there seems to be no way to create an icon for the .fap in the flipper zero, is this planned or is there a way to do this and i'm just dumb?

**Comment by dcoles** [comment]
This isn't currently supported by the [`manifest!`](https://github.com/dcoles/flipperzero-rs/blob/cf4725678e6d3b5a9790aac1a3630a4f88d0bec8/crates/rt/src/manifest.rs#L12) macro, but support can definitely be added (see #21).

I think the only real "trick" is converting an image to the format expected by the Flipper Zero. The two options would be:
- Have the user pre-build the icon and just include it in the binary (what we do in the [images example](https://github.com/dcoles/flipperzero-rs/blob/cf4725678e6d3b5a9790aac1a3630a4f88d0bec8/crates/flipperzero/examples/example_images.rs))
- Use `build.rs` or a proc macro to process the image when creating the manifest

I'm probably learning towards the later, since it would make a better experience for developers.

**Reply by GitCat3** [reply]
Thanks for the info! Quick question, is any of the code in the sample that you linked actually required to set the file as the icon or can I just include it?

**Reply by dcoles** [reply]
I *think* icons used for an external application have the same format as the larger ones in that example. However I believe the icon data needs to be [stored in the manifest struct](https://github.com/dcoles/flipperzero-rs/blob/cf4725678e6d3b5a9790aac1a3630a4f88d0bec8/crates/rt/src/manifest.rs#L71) rather than just being a pointer to some other data in the binary.

**Reply by GitCat3** [reply]
I am pretty new to rust, I see a thing in the manifest file that you linked that says hasIcon and the icon itself, so do I do something with the manifest in my project to make it include that?

**Reply by dcoles** [reply]
Up until today, it wasn't actually possible to set those fields using the [`manifest!` macro](https://docs.rs/flipperzero-rt/0.4.1/flipperzero_rt/macro.manifest.html).

This is now fixed in `0.4.1` (see [updated answer](https://github.com/dcoles/flipperzero-rs/discussions/20#discussioncomment-4362840)).

**Comment by dcoles** [comment, marked as answer]
As of [`flipperzero-rt-0.4.1`](https://crates.io/crates/flipperzero-rt/0.4.1) it's now possible to set a custom icon for your application! 🥳 

**From [`examples/hello-rust`](https://github.com/dcoles/flipperzero-rs/blob/main/examples/hello-rust/src/main.rs#L16-L18):**
```rust
// Define the FAP Manifest for this application
manifest!(
    name = "Hello, Rust!",
    has_icon = true,
    // See `docs/icons.md` for icon format
    icon = "rustacean-10x10.icon"
);
```

Icons can be created from any 10x10 px monochrome image. However they need to be converted to the [Flipper Zero's custom icon format](/dcoles/flipperzero-rs/blob/main/docs/icons.md) before they can be used.

**Reply by GitCat3** [reply]
Wow thanks!
